File: packages/uns-core/src/uns-config/secret-resolver.ts
import os from "node:os";
import type { NetworkInterfaceInfo } from "node:os";
import type { AppConfig } from "../app-config";
import {
  isSecretPlaceholder,
  type EnvSecretPlaceholder,
  type InfisicalSecretPlaceholder,
  type SecretPlaceholder,
} from "./secret-placeholders";
import {
  isHostPlaceholder,
  type HostPlaceholder,
  type SystemHostPlaceholder,
} from "./host-placeholders";

export type SecretValueResolved = string | undefined;
export type HostValueResolved = string | undefined;

type SecretPlaceholderCandidate =
  | { provider: "env"; key: string }
  | { provider: "infisical"; key: string; path: string };

type HostPlaceholderCandidate = HostPlaceholder;

export type ResolvedConfig<T> = T extends SecretPlaceholderCandidate
  ? SecretValueResolved
  : T extends HostPlaceholderCandidate
    ? HostValueResolved
  : T extends (infer U)[]
    ? ResolvedConfig<U>[]
    : T extends Record<string, unknown>
      ? { [K in keyof T]: ResolvedConfig<T[K]> }
      : T;

export type ResolvedAppConfig = ResolvedConfig<AppConfig>;

export interface InfisicalFetchRequest {
  path: string;
  key: string;
  environment?: string;
  projectId?: string;
  type?: "shared" | "personal";
}

export type InfisicalFetcher = (
  request: InfisicalFetchRequest
) => Promise<string | undefined>;

export interface InfisicalResolverOptions {
  /**
   * Provide a custom fetcher. If omitted, the resolver tries to lazily instantiate
   * an Infisical client using @infisical/sdk based on the supplied token/siteUrl/options.
   */
  fetchSecret?: InfisicalFetcher;
  /**
   * Machine token or personal token used when creating the default Infisical client.
   * Falls back to the INFISICAL_TOKEN or INFISICAL_PERSONAL_TOKEN environment variables.
   */
  token?: string;
  /** Optional Infisical site URL override. Defaults to INFISICAL_SITE_URL when present. */
  siteUrl?: string;
  /** Default environment used when a placeholder does not specify one explicitly. */
  environment?: string;
  /** Default project id used when a placeholder does not provide one. */
  projectId?: string;
  /** Default secret type. Shared secrets are used by default. */
  type?: "shared" | "personal";
  /** Disable in-memory caching when set to false. Enabled by default. */
  cache?: boolean;
}

export interface SecretResolverOptions {
  /** Environment map used for `env` placeholders. Defaults to process.env. */
  env?: NodeJS.ProcessEnv;
  /** Configuration for resolving Infisical placeholders. */
  infisical?: InfisicalResolverOptions;
  /** Callback invoked before throwing when a required secret cannot be resolved. */
  onMissingSecret?: (placeholder: SecretPlaceholder, source: "env" | "infisical") => void;
  /** Configuration for resolving host placeholders. */
  hosts?: HostResolverOptions;
}

export interface HostResolverOptions {
  /** Optional environment map used when falling back to process.env lookups. */
  env?: NodeJS.ProcessEnv;
  /** Static mapping of external host keys to concrete host strings. */
  externalHosts?: Record<string, string | undefined>;
  /**
   * Custom resolver invoked when an external host needs to be resolved.
   * It can return synchronously or asynchronously.
   */
  resolveExternal?: (key: string) => string | undefined | Promise<string | undefined>;
  /** Override for os.networkInterfaces (handy for tests). */
  networkInterfaces?: () => Record<string, NetworkInterfaceInfo[] | undefined>;
  /** Callback invoked before throwing when a required host cannot be resolved. */
  onMissingHost?: (placeholder: HostPlaceholder) => void;
}

const envCache = new Map<string, string | undefined>();
const infisicalCache = new Map<string, Promise<string | undefined>>();

const structuredCloneFallback = <T>(value: T): T =>
  typeof structuredClone === "function"
    ? structuredClone(value)
    : (JSON.parse(JSON.stringify(value)) as T);

export async function resolveConfigSecrets(
  config: AppConfig,
  options: SecretResolverOptions = {}
): Promise<ResolvedAppConfig> {
  const working = structuredCloneFallback(config);
  await resolveNode(working, options);
  return working as ResolvedAppConfig;
}

export function clearSecretResolverCaches(): void {
  envCache.clear();
  infisicalCache.clear();
  defaultInfisicalFetcherCache.clear();
}

async function resolveNode(node: unknown, options: SecretResolverOptions): Promise<unknown> {
  if (isSecretPlaceholder(node)) {
    return resolveSecretValue(node, options);
  }

  if (isHostPlaceholder(node)) {
    return resolveHostValue(node, options.hosts);
  }

  if (Array.isArray(node)) {
    const resolvedArray = await Promise.all(node.map(item => resolveNode(item, options)));
    return resolvedArray;
  }

  if (node && typeof node === "object") {
    const obj = node as Record<string, unknown>;

    for (const key of Object.keys(obj)) {
      const current = obj[key];

      if (isSecretPlaceholder(current)) {
        const resolved = await resolveSecretValue(current, options);
        if (resolved === undefined) {
          delete obj[key];
        } else {
          obj[key] = resolved;
        }
        continue;
      }

      if (isHostPlaceholder(current)) {
        const resolved = await resolveHostValue(current, options.hosts);
        if (resolved === undefined) {
          delete obj[key];
        } else {
          obj[key] = resolved;
        }
        continue;
      }

      const resolvedChild = await resolveNode(current, options);
      obj[key] = resolvedChild as unknown;
    }

    return obj;
  }

  return node;
}

async function resolveSecretValue(
  placeholder: SecretPlaceholder,
  options: SecretResolverOptions
): Promise<SecretValueResolved> {
  switch (placeholder.provider) {
    case "env":
      return resolveEnvSecret(placeholder, options);
    case "infisical":
      return resolveInfisicalSecret(placeholder, options);
    default:
      // Exhaustive check to guard future provider additions.
      return assertNeverProvider(placeholder);
  }
}

function assertNeverProvider(x: never): never {
  throw new Error(`Unsupported secret provider: ${JSON.stringify(x)}`);
}

async function resolveHostValue(
  placeholder: HostPlaceholder,
  options?: HostResolverOptions
): Promise<HostValueResolved> {
  const effectiveOptions = options ?? {};

  switch (placeholder.provider) {
    case "inline":
      return placeholder.value;
    case "external": {
      const envMap = effectiveOptions.env ?? process.env;

      let resolved = await maybeResolveExternal(
        placeholder.key,
        effectiveOptions.resolveExternal
      );

      if (resolved === undefined) {
        resolved = effectiveOptions.externalHosts?.[placeholder.key];
      }

      if (resolved === undefined) {
        resolved = envMap?.[placeholder.key];
      }

      if (resolved === undefined) {
        if (placeholder.default !== undefined) {
          return placeholder.default;
        }
        if (placeholder.optional) {
          return undefined;
        }

        effectiveOptions.onMissingHost?.(placeholder);
        throw new Error(`External host '${placeholder.key}' could not be resolved.`);
      }

      return resolved;
    }
    case "system":
      return resolveSystemHost(placeholder, effectiveOptions);
    default:
      return assertNeverHostProvider(placeholder);
  }
}

async function maybeResolveExternal(
  key: string,
  resolver?: (key: string) => string | undefined | Promise<string | undefined>
): Promise<string | undefined> {
  if (!resolver) {
    return undefined;
  }

  try {
    return await resolver(key);
  } catch (error) {
    throw new Error(
      `Failed to resolve external host '${key}' via custom resolver: ${
        error instanceof Error ? error.message : String(error)
      }`,
      { cause: error }
    );
  }
}

function assertNeverHostProvider(x: never): never {
  throw new Error(`Unsupported host provider: ${JSON.stringify(x)}`);
}

function resolveSystemHost(
  placeholder: SystemHostPlaceholder,
  options: HostResolverOptions
): HostValueResolved {
  const getInterfaces = options.networkInterfaces ?? os.networkInterfaces;
  const interfaces = getInterfaces();
  const family = placeholder.family ?? "IPv4";
  const targetName = placeholder.interfaceName;

  const interfaceEntries: Array<readonly [string, NetworkInterfaceInfo[] | undefined]> = targetName
    ? [[targetName, interfaces[targetName]]]
    : Object.entries(interfaces);

  for (const [name, ifaceList] of interfaceEntries) {
    if (!ifaceList) {
      continue;
    }
    for (const iface of ifaceList) {
      if (
        iface &&
        iface.family === family &&
        iface.internal !== true &&
        typeof iface.address === "string" &&
        iface.address.length > 0
      ) {
        return iface.address;
      }
    }

    if (targetName) {
      break;
    }
  }

  if (placeholder.default !== undefined) {
    return placeholder.default;
  }

  if (placeholder.optional) {
    return undefined;
  }

  options.onMissingHost?.(placeholder);
  const targetDescription = targetName ? `interface '${targetName}'` : "available interfaces";
  throw new Error(
    `System host lookup failed for ${targetDescription} (family ${family}).`
  );
}

async function resolveEnvSecret(
  placeholder: EnvSecretPlaceholder,
  options: SecretResolverOptions
): Promise<SecretValueResolved> {
  const envMap = options.env ?? process.env;
  const cacheKey = placeholder.key;

  if (envCache.has(cacheKey)) {
    return envCache.get(cacheKey) ?? undefined;
  }

  const value = envMap[cacheKey];

  if (value === undefined || value === null) {
    if (placeholder.default !== undefined) {
      envCache.set(cacheKey, placeholder.default);
      return placeholder.default;
    }
    if (placeholder.optional) {
      envCache.set(cacheKey, undefined);
      return undefined;
    }

    options.onMissingSecret?.(placeholder, "env");
    throw new Error(`Required environment variable '${cacheKey}' is not set.`);
  }

  envCache.set(cacheKey, value);
  return value;
}

async function resolveInfisicalSecret(
  placeholder: InfisicalSecretPlaceholder,
  options: SecretResolverOptions
): Promise<SecretValueResolved> {
  const infisicalOptions = options.infisical;
  const fetcher = await getInfisicalFetcher(infisicalOptions);

  if (!fetcher) {
    options.onMissingSecret?.(placeholder, "infisical");
    throw new Error(
      "Infisical secret requested but no resolver is configured. Provide SecretResolverOptions.infisical."
    );
  }

  const environment = placeholder.environment ?? infisicalOptions?.environment ?? process.env.INFISICAL_ENVIRONMENT;
  const projectId = placeholder.projectId ?? infisicalOptions?.projectId ?? process.env.INFISICAL_PROJECT_ID;
  const type = infisicalOptions?.type ?? "shared";

  if (!environment) {
    throw new Error(
      `Infisical secret '${placeholder.path}:${placeholder.key}' is missing an environment. ` +
      "Set it on the placeholder, pass SecretResolverOptions.infisical.environment, or define INFISICAL_ENVIRONMENT."
    );
  }

  if (!projectId) {
    throw new Error(
      `Infisical secret '${placeholder.path}:${placeholder.key}' is missing a project id. ` +
      "Set it on the placeholder, pass SecretResolverOptions.infisical.projectId, or define INFISICAL_PROJECT_ID."
    );
  }

  const cacheKey = JSON.stringify({
    path: placeholder.path,
    key: placeholder.key,
    environment,
    projectId,
    type,
  });

  if (infisicalOptions?.cache !== false && infisicalCache.has(cacheKey)) {
    return infisicalCache.get(cacheKey) ?? undefined;
  }

  const fetchPromise = fetcher({
    path: placeholder.path,
    key: placeholder.key,
    environment,
    projectId,
    type,
  }).then(secret => {
    if (secret === undefined || secret === null) {
      if (placeholder.default !== undefined) {
        return placeholder.default;
      }
      if (placeholder.optional) {
        return undefined;
      }
      options.onMissingSecret?.(placeholder, "infisical");
      throw new Error(
        `Secret '${placeholder.path}:${placeholder.key}' not found in Infisical.`
      );
    }
    return secret;
  });

  if (infisicalOptions?.cache !== false) {
    infisicalCache.set(cacheKey, fetchPromise);
  }

  return fetchPromise;
}

async function getInfisicalFetcher(
  options?: InfisicalResolverOptions
): Promise<InfisicalFetcher | undefined> {
  if (!options) {
    return undefined;
  }

  if (options.fetchSecret) {
    return options.fetchSecret;
  }

  const token =
    options.token ??
    process.env.INFISICAL_TOKEN ??
    process.env.INFISICAL_PERSONAL_TOKEN ??
    "";

  if (!token) {
    return undefined;
  }

  const siteUrlKey = options.siteUrl ?? process.env.INFISICAL_SITE_URL ?? "";
  const cacheKey = `${token}::${siteUrlKey}`;

  if (!defaultInfisicalFetcherCache.has(cacheKey)) {
    defaultInfisicalFetcherCache.set(
      cacheKey,
      createDefaultInfisicalFetcher(token, options.siteUrl)
    );
  }

  const baseFetcher = await defaultInfisicalFetcherCache.get(cacheKey)!;
  return baseFetcher;
}

const defaultInfisicalFetcherCache = new Map<string, Promise<InfisicalFetcher>>();

async function createDefaultInfisicalFetcher(
  token: string,
  siteUrl?: string
): Promise<InfisicalFetcher> {
  try {
    const sdkModule: typeof import("@infisical/sdk") = await import("@infisical/sdk");
    const { InfisicalSDK, SecretType } = sdkModule;
    if (!InfisicalSDK) {
      throw new Error("@infisical/sdk does not export InfisicalSDK");
    }

    const sdkInstance = new InfisicalSDK({ siteUrl });
    const authenticatedSdk = sdkInstance.auth().accessToken(token);

    return async ({ path, key, environment, projectId, type = "shared" }: InfisicalFetchRequest) => {
      if (!environment) {
        throw new Error(
          `Infisical secret '${path}:${key}' is missing an environment.`
        );
      }
      if (!projectId) {
        throw new Error(
          `Infisical secret '${path}:${key}' is missing a project id.`
        );
      }

      const secret = await authenticatedSdk
        .secrets()
        .getSecret({
          secretName: key,
          secretPath: path,
          environment,
          projectId,
          viewSecretValue: true,
          type: type === "personal" ? SecretType.Personal : SecretType.Shared,
        });

      return secret?.secretValue ?? undefined;
    };
  } catch (error) {
    throw new Error(
      "Failed to initialize @infisical/sdk. Install the package or provide SecretResolverOptions.infisical.fetchSecret.",
      { cause: error as Error }
    );
  }
}

File: packages/uns-py/src/uns_kit/core/secret_resolver.py
from __future__ import annotations

import copy
import os
import socket
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable, Dict, Mapping, Optional, Sequence

INFISICAL_TOKEN_FILE = "/run/secrets/infisical_token"
INFISICAL_PROJECT_ID_FILE = "/run/secrets/infisical_project_id"
INFISICAL_SITE_URL_FILE = "/run/secrets/infisical_site_url"
INFISICAL_TOKEN_FILE_ALT = "/var/lib/uns/secrets/infisical_token"
INFISICAL_PROJECT_ID_FILE_ALT = "/var/lib/uns/secrets/infisical_project_id"
INFISICAL_SITE_URL_FILE_ALT = "/var/lib/uns/secrets/infisical_site_url"


SecretFetcher = Callable[[str, str, str, str, str], Optional[str]]
ExternalHostResolver = Callable[[str], Optional[str]]
MissingSecretCallback = Callable[[Mapping[str, Any], str], None]
MissingHostCallback = Callable[[Mapping[str, Any]], None]


@dataclass
class InfisicalResolverOptions:
    fetch_secret: Optional[SecretFetcher] = None
    token: Optional[str] = None
    site_url: Optional[str] = None
    environment: Optional[str] = None
    project_id: Optional[str] = None
    secret_type: str = "shared"
    cache: bool = True


@dataclass
class HostResolverOptions:
    env: Optional[Mapping[str, str]] = None
    external_hosts: Optional[Mapping[str, Optional[str]]] = None
    resolve_external: Optional[ExternalHostResolver] = None
    on_missing_host: Optional[MissingHostCallback] = None


@dataclass
class SecretResolverOptions:
    env: Optional[Mapping[str, str]] = None
    infisical: Optional[InfisicalResolverOptions] = None
    on_missing_secret: Optional[MissingSecretCallback] = None
    hosts: Optional[HostResolverOptions] = None


_env_cache: Dict[str, Optional[str]] = {}
_infisical_cache: Dict[str, Optional[str]] = {}
_default_infisical_fetcher_cache: Dict[str, SecretFetcher] = {}


def clear_secret_resolver_caches() -> None:
    _env_cache.clear()
    _infisical_cache.clear()
    _default_infisical_fetcher_cache.clear()


def resolve_infisical_config(
    options: Optional[InfisicalResolverOptions] = None,
) -> Dict[str, Optional[str]]:
    return {
        "token": _resolve_infisical_token(options),
        "projectId": _resolve_infisical_project_id(options),
        "siteUrl": _resolve_infisical_site_url(options),
    }


def resolve_config_secrets(
    config: Mapping[str, Any],
    options: Optional[SecretResolverOptions] = None,
) -> Dict[str, Any]:
    effective_options = options or SecretResolverOptions()
    working = copy.deepcopy(dict(config))
    return _resolve_node(working, effective_options)


def _resolve_node(node: Any, options: SecretResolverOptions) -> Any:
    if _is_secret_placeholder(node):
        return _resolve_secret_value(node, options)

    if _is_host_placeholder(node):
        return _resolve_host_value(node, options.hosts)

    if isinstance(node, list):
        return [_resolve_node(item, options) for item in node]

    if isinstance(node, dict):
        keys = list(node.keys())
        for key in keys:
            current = node[key]
            if _is_secret_placeholder(current):
                resolved = _resolve_secret_value(current, options)
                if resolved is None:
                    del node[key]
                else:
                    node[key] = resolved
                continue

            if _is_host_placeholder(current):
                resolved = _resolve_host_value(current, options.hosts)
                if resolved is None:
                    del node[key]
                else:
                    node[key] = resolved
                continue

            node[key] = _resolve_node(current, options)
        return node

    return node


def _is_secret_placeholder(value: Any) -> bool:
    return (
        isinstance(value, dict)
        and value.get("provider") in {"env", "infisical"}
        and isinstance(value.get("key"), str)
    )


def _is_host_placeholder(value: Any) -> bool:
    if not isinstance(value, dict):
        return False
    provider = value.get("provider")
    return provider in {"inline", "external", "system"}


def _resolve_secret_value(
    placeholder: Mapping[str, Any],
    options: SecretResolverOptions,
) -> Optional[str]:
    provider = placeholder.get("provider")
    if provider == "env":
        return _resolve_env_secret(placeholder, options)
    if provider == "infisical":
        return _resolve_infisical_secret(placeholder, options)
    raise ValueError(f"Unsupported secret provider: {provider!r}")


def _resolve_host_value(
    placeholder: Mapping[str, Any],
    options: Optional[HostResolverOptions],
) -> Optional[str]:
    effective_options = options or HostResolverOptions()
    provider = placeholder.get("provider")

    if provider == "inline":
        value = placeholder.get("value")
        if isinstance(value, str) and value:
            return value
        raise ValueError("Inline host placeholder requires a non-empty 'value'.")

    if provider == "external":
        key = placeholder.get("key")
        if not isinstance(key, str) or not key:
            raise ValueError("External host placeholder requires a non-empty 'key'.")

        resolved: Optional[str] = None
        if effective_options.resolve_external:
            resolved = effective_options.resolve_external(key)
        if resolved is None and effective_options.external_hosts:
            resolved = effective_options.external_hosts.get(key)
        if resolved is None:
            env_map = effective_options.env or os.environ
            resolved = env_map.get(key)

        if resolved is None:
            default = placeholder.get("default")
            optional = bool(placeholder.get("optional"))
            if isinstance(default, str):
                return default
            if optional:
                return None
            if effective_options.on_missing_host:
                effective_options.on_missing_host(placeholder)
            raise ValueError(f"External host '{key}' could not be resolved.")
        return resolved

    if provider == "system":
        return _resolve_system_host(placeholder, effective_options)

    raise ValueError(f"Unsupported host provider: {provider!r}")


def _resolve_system_host(
    placeholder: Mapping[str, Any],
    options: HostResolverOptions,
) -> Optional[str]:
    family = placeholder.get("family") or "IPv4"
    interface_name = placeholder.get("interfaceName")
    optional = bool(placeholder.get("optional"))
    default = placeholder.get("default")

    resolved: Optional[str] = None
    try:
        if family == "IPv6":
            # If no interface callback exists, this best-effort fallback mirrors
            # TS behavior of returning the first non-empty non-loopback address.
            s = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
            s.connect(("2001:4860:4860::8888", 80))
            resolved = s.getsockname()[0]
            s.close()
        else:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(("8.8.8.8", 80))
            resolved = s.getsockname()[0]
            s.close()
    except Exception:
        resolved = None

    if resolved:
        return resolved

    if isinstance(default, str):
        return default
    if optional:
        return None
    if options.on_missing_host:
        options.on_missing_host(placeholder)
    target = f"interface '{interface_name}'" if isinstance(interface_name, str) else "available interfaces"
    raise ValueError(f"System host lookup failed for {target} (family {family}).")


def _resolve_env_secret(
    placeholder: Mapping[str, Any],
    options: SecretResolverOptions,
) -> Optional[str]:
    key = placeholder.get("key")
    if not isinstance(key, str) or not key:
        raise ValueError("env placeholder requires a non-empty 'key'.")

    if key in _env_cache:
        return _env_cache[key]

    env_map = options.env or os.environ
    value = env_map.get(key)

    if value is None:
        default = placeholder.get("default")
        optional = bool(placeholder.get("optional"))
        if isinstance(default, str):
            _env_cache[key] = default
            return default
        if optional:
            _env_cache[key] = None
            return None
        if options.on_missing_secret:
            options.on_missing_secret(placeholder, "env")
        raise ValueError(f"Required environment variable '{key}' is not set.")

    _env_cache[key] = value
    return value


def _resolve_infisical_secret(
    placeholder: Mapping[str, Any],
    options: SecretResolverOptions,
) -> Optional[str]:
    infisical_options = options.infisical or InfisicalResolverOptions()
    fetcher = _get_infisical_fetcher(infisical_options)

    path = placeholder.get("path")
    key = placeholder.get("key")
    if not isinstance(path, str) or not path:
        raise ValueError("infisical placeholder requires a non-empty 'path'.")
    if not isinstance(key, str) or not key:
        raise ValueError("infisical placeholder requires a non-empty 'key'.")

    environment = (
        placeholder.get("environment")
        or infisical_options.environment
        or os.environ.get("INFISICAL_ENVIRONMENT")
    )
    project_id = placeholder.get("projectId") or _resolve_infisical_project_id(infisical_options)
    secret_type = infisical_options.secret_type or "shared"

    cache_key = f"{path}|{key}|{environment}|{project_id}|{secret_type}"
    use_cache = infisical_options.cache is not False
    cached = _infisical_cache.get(cache_key) if use_cache else None

    if fetcher is None:
        if cached is not None:
            print(
                f"Infisical fetcher unavailable; using cached secret for {path}:{key}."
            )
            return cached
        fallback = _fallback_placeholder_value(placeholder)
        print(
            f"Infisical fetcher unavailable; returning "
            f"{'undefined' if fallback is None else 'default'} for {path}:{key}."
        )
        if fallback is None and not bool(placeholder.get("optional")) and options.on_missing_secret:
            options.on_missing_secret(placeholder, "infisical")
        return fallback

    if not isinstance(environment, str) or not environment:
        raise ValueError(
            f"Infisical secret '{path}:{key}' is missing an environment. "
            "Set it on the placeholder, pass InfisicalResolverOptions.environment, "
            "or define INFISICAL_ENVIRONMENT."
        )

    if not isinstance(project_id, str) or not project_id:
        raise ValueError(
            f"Infisical secret '{path}:{key}' is missing a project id. "
            "Set it on the placeholder, pass InfisicalResolverOptions.project_id, "
            "define INFISICAL_PROJECT_ID, or provide /run/secrets/infisical_project_id "
            "(also /var/lib/uns/secrets/infisical_project_id)."
        )

    if cached is not None:
        return cached

    try:
        secret = fetcher(path, key, environment, project_id, secret_type)
        if secret is None:
            fallback = _fallback_placeholder_value(placeholder)
            if fallback is None and not bool(placeholder.get("optional")):
                if options.on_missing_secret:
                    options.on_missing_secret(placeholder, "infisical")
                raise ValueError(f"Secret '{path}:{key}' not found in Infisical.")
            secret = fallback
    except Exception as error:
        if cached is not None:
            print(
                f"Infisical fetch failed ({error}); using cached secret for {path}:{key}."
            )
            return cached

        fallback = _fallback_placeholder_value(placeholder)
        print(
            f"Infisical fetch failed ({error}); returning "
            f"{'undefined' if fallback is None else 'default'} for {path}:{key}."
        )
        if fallback is None and not bool(placeholder.get("optional")):
            if options.on_missing_secret:
                options.on_missing_secret(placeholder, "infisical")
            raise ValueError(
                f"Failed to fetch Infisical secret {path}:{key}: {error}"
            ) from error
        secret = fallback

    if use_cache:
        _infisical_cache[cache_key] = secret
    return secret


def _fallback_placeholder_value(placeholder: Mapping[str, Any]) -> Optional[str]:
    default = placeholder.get("default")
    if isinstance(default, str):
        return default
    if bool(placeholder.get("optional")):
        return None
    return None


def _get_infisical_fetcher(
    options: InfisicalResolverOptions,
) -> Optional[SecretFetcher]:
    if options.fetch_secret:
        return options.fetch_secret

    token = _resolve_infisical_token(options)
    if not token:
        return None

    site_url = _resolve_infisical_site_url(options) or "https://app.infisical.com"
    cache_key = f"{token}::{site_url}"
    if cache_key in _default_infisical_fetcher_cache:
        return _default_infisical_fetcher_cache[cache_key]

    fetcher = _create_default_infisical_fetcher(token, site_url)
    _default_infisical_fetcher_cache[cache_key] = fetcher
    return fetcher


def _create_default_infisical_fetcher(token: str, site_url: str) -> SecretFetcher:
    try:
        from infisical_sdk import InfisicalSDKClient  # type: ignore
    except Exception as error:
        raise ValueError(
            "Failed to initialize Infisical SDK. Install 'infisicalsdk' or provide "
            "InfisicalResolverOptions.fetch_secret."
        ) from error

    client = InfisicalSDKClient(host=site_url, token=token)

    def fetch(path: str, key: str, environment: str, project_id: str, _type: str) -> Optional[str]:
        # Type is currently accepted for parity with uns-core options.
        response = client.secrets.get_secret_by_name(
            secret_name=key,
            project_id=project_id,
            environment_slug=environment,
            secret_path=path,
            include_imports=True,
            expand_secret_references=True,
            view_secret_value=True,
        )

        secret_value = _extract_secret_value(response)
        if secret_value is None:
            return None
        return str(secret_value)

    return fetch


def _extract_secret_value(response: Any) -> Optional[str]:
    candidates: Sequence[str] = ("secretValue", "secret_value", "value")
    for field in candidates:
        if hasattr(response, field):
            value = getattr(response, field)
            if value is not None:
                return str(value)

    if isinstance(response, Mapping):
        for field in candidates:
            value = response.get(field)
            if value is not None:
                return str(value)

    # Some SDK responses wrap the secret object in "secret".
    if hasattr(response, "secret"):
        wrapped = getattr(response, "secret")
        return _extract_secret_value(wrapped)
    if isinstance(response, Mapping) and isinstance(response.get("secret"), Mapping):
        return _extract_secret_value(response.get("secret"))
    return None


def _resolve_infisical_token(options: Optional[InfisicalResolverOptions]) -> Optional[str]:
    candidate = (
        (options.token if options else None)
        or os.environ.get("INFISICAL_TOKEN")
        or os.environ.get("INFISICAL_PERSONAL_TOKEN")
        or _read_secret_file(INFISICAL_TOKEN_FILE)
        or _read_secret_file(INFISICAL_TOKEN_FILE_ALT)
    )
    return _normalize_secret_value(candidate)


def _resolve_infisical_site_url(options: Optional[InfisicalResolverOptions]) -> Optional[str]:
    candidate = (
        (options.site_url if options else None)
        or os.environ.get("INFISICAL_SITE_URL")
        or _read_secret_file(INFISICAL_SITE_URL_FILE)
        or _read_secret_file(INFISICAL_SITE_URL_FILE_ALT)
    )
    return _normalize_secret_value(candidate)


def _resolve_infisical_project_id(options: Optional[InfisicalResolverOptions]) -> Optional[str]:
    candidate = (
        (options.project_id if options else None)
        or os.environ.get("INFISICAL_PROJECT_ID")
        or _read_secret_file(INFISICAL_PROJECT_ID_FILE)
        or _read_secret_file(INFISICAL_PROJECT_ID_FILE_ALT)
    )
    return _normalize_secret_value(candidate)


def _normalize_secret_value(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed if trimmed else None


def _read_secret_file(file_path: str) -> Optional[str]:
    p = Path(file_path)
    if not p.exists():
        return None
    try:
        return p.read_text(encoding="utf-8")
    except Exception as error:
        raise ValueError(
            f"Failed to read Infisical secret file '{file_path}': {error}"
        ) from error
